### PR TITLE
Use new-billing API output format (used in statement, report)

### PR DESCRIPTION
**Pull request by schmie**
What
----

- Introduce data type and parsers to handle new-billing output for statements and reports.
- Switch over data type and parser use to actually work against the new output (this will require additional steps to minimise downtime [link](https://hackmd.cloudapps.digital/bR_sdMLASVSbozHGbUyLfQ)).
- Change time range stop handling in current month to explicitly pass current time (new-billing would otherwise extrapolate data).

How to review
-------------

- Code review
- Deploy/test against dev env (we currently have dfe production data available in `dev03` in org `mini-test`).
- Testing might require to point the BILLING_URL environment variable of paas-admin against a new-billing endpoint (e.g.: `cf set-env paas-admin BILLING_URL https://new-billing.dev03.dev.cloudpipeline.digital; cf restage paas-admin;`).
I (@schmie) am happy to pair if anything is unclear : )

Who can review
---------------

not @schmie 

---

🚨⚠️ Please do not merge this pull request via the GitHub UI ⚠️🚨

### CODE REVIEW

File: src/components/statements/controllers.tsx [hunk 1]
```diff
@@ -202,6 +202,15 @@ export async function viewStatement(
...
   }
 
+  const today = new Date();
+
+  let rangeStop;
+  if (format(startOfMonth(today), YYYMMDD) == format(startOfMonth(rangeStart), YYYMMDD)) {
+    rangeStop = format(today, "yyyy-MM-dd'T'HH:mm:ss'Z'");
+  } else {
+    rangeStop = add(rangeStart, { months: 1 });
+  }
```
Comment (kr8n3r) [inline src/components/statements/controllers.tsx:212]: can you explain what we're trying to do here?

Comment (kr8n3r) [reply src/components/statements/controllers.tsx:212]: thanks for explaining. just change `==` to `===`

Comment (kr8n3r) [reply src/components/statements/controllers.tsx:212]: @schmie if you take away the `ormat`,could you use https://date-fns.org/v2.28.0/docs/isEqual to compare dates?

**Comment by kr8n3r** [comment]
here's what the output is for the failing test at https://github.com/alphagov/paas-admin/runs/5112056845?check_suite_focus=true#step:6:2175


<html>
<body>
<h1 class="govuk-heading-l" style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Billables for February 2022</h1><div class="govuk-grid-row" style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="govuk-grid-column-one-quarter"><h2 class="govuk-heading-m">1<span> </span><span class="govuk-caption-m">Billable events</span></h2></div><div class="govuk-grid-column-one-quarter"><h2 class="govuk-heading-m">£0.02<span> </span><span class="govuk-caption-m">Total including VAT</span></h2></div><div class="govuk-grid-column-one-quarter"><h2 class="govuk-heading-m">£0.02<span> </span><span class="govuk-caption-m">Total excluding VAT</span></h2></div><div class="govuk-grid-column-one-quarter"><h2 class="govuk-heading-m">£0.02<span> </span><span class="govuk-caption-m">Total excluding VAT including fee</span></h2></div></div><h1 class="govuk-heading-l" style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Billables by organisation for February 2022</h1><div class="scrollable-table-container" style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Organization | Quota | Including VAT | Excluding VAT | Excluding VAT including fee
-- | -- | -- | -- | --
name-1996 | £0.00 | £0.00 | £0.00

</div>
</body>
</html>

whereas on main the output is

<html>
<body>
<h1 class="govuk-heading-l" style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Billables for February 2022</h1><div class="govuk-grid-row" style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="govuk-grid-column-one-quarter"><h2 class="govuk-heading-m">1<span> </span><span class="govuk-caption-m">Billable events</span></h2></div><div class="govuk-grid-column-one-quarter"><h2 class="govuk-heading-m">£0.02<span> </span><span class="govuk-caption-m">Total including VAT</span></h2></div><div class="govuk-grid-column-one-quarter"><h2 class="govuk-heading-m">£0.02<span> </span><span class="govuk-caption-m">Total excluding VAT</span></h2></div><div class="govuk-grid-column-one-quarter"><h2 class="govuk-heading-m">£0.02<span> </span><span class="govuk-caption-m">Total excluding VAT including fee</span></h2></div></div><h1 class="govuk-heading-l" style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Billables by organisation for February 2022</h1><div class="scrollable-table-container" style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Organization | Quota | Including VAT | Excluding VAT | Excluding VAT including fee
-- | -- | -- | -- | --
name-1996 | £0.02 | £0.02 | £0.02

</div>
</body>
</html>

**Comment by kr8n3r** [comment]
@schmie 
deployed to dev03 and get the same error as with local dev


```
":10,"message":"accounts received 404 when making get request to /users/99022be6-feb8-4f78-96f3-7d11f4d476f1 in 10ms"}
{"level":50,"time":1645096658933,"pid":15839,"hostname":"GDS7463","req":{"method":"GET","url":"/organisations/a7aff246-5f5b-4cf8-87d8-f316053e4a20/statements/2022-02-01"},"stack":"Error: Billing is currently unavailable, please try again later.\n    at Object.<anonymous> (/Users/janikraner/paas/paas-admin/dist/webpack:/paas-admin/src/components/statements/controllers.tsx:281:11)\n    at Generator.throw (<anonymous>)\n    at rejected (/Users/janikraner/paas/paas-admin/node_modules/tslib/tslib.js:115:69)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)","type":"Error","msg":"Billing is currently unavailable, please try again later."}
```

Screenshot 2022-02-17 at 11 27 53

File: src/lib/billing/types.ts [hunk 2]
```diff
@@ -110,6 +110,23 @@ export interface IBillableEventResponse extends IUsageEventResponse {
...
+  readonly plan_name: string;
+  readonly space_guid: string;
+  readonly space_name: string;
+  readonly resource_guid: string;
+  readonly resource_type: string;
+  readonly resource_name: string;
+  readonly component_name: string;
+  readonly charge_gbp_inc_vat: string;
+  readonly charge_gbp_exc_vat: string;
+  readonly charge_usd_exc_vat: string;
```
Comment (kr8n3r) [inline src/lib/billing/types.ts:127]: can't find any docs on new billing, so just checking if they're meant to be string and not Number?

**Comment by schmie** [comment]
Hi @kr8n3r
Thanks for reviewing!
The step `cf set-env paas-admin BILLING_URL https://new-billing.dev03.dev.cloudpipeline.digital; cf restage paas-admin;` was missing - ran it in `dev03`, if you like to have another look.
Cheers!

**Comment by kr8n3r** [comment]
@schmie  thanks
org billing and calculator are working but admin overall cost report isn't working https://admin.dev03.dev.cloudpipeline.digital/reports/cost/2022-01-01

**Comment by whpearson** [comment]
> @schmie thanks org billing and calculator are working but admin overall cost report isn't working https://admin.dev03.dev.cloudpipeline.digital/reports/cost/2022-01-01

We think this is because we bought some production data into the dev environment (and the quota GUIDs are different). I'll discuss with seb the best way to test this.

**Comment by schmie** [comment]
We updated the data and reports are working, now.

**Review by kr8n3r — APPROVED** [review]
code looks good to me, though i can't verify the truthiness of provided source data

I'm assuming we'll do old code cleanup at a later date, once we're happy new API is solid

**Review by kr8n3r — COMMENTED** [review]
I'll leave it to you to merge and do whatever needs to happen for it to make it to prod

**Pull request closed without merging**